File: backend/semgrep_rules/tests/javascript/kisa_sw_03_path_traversal.js
const fs =
  require("fs");

const path =
  require("path");


function unsafeDirectRead(req) {
  const filename =
    req.query.file;

  // ruleid: kisa.sw03.javascript.external-input-to-file-path
  return fs.readFileSync(filename, "utf8");
}


function unsafeJoin(req, base) {
  const filename =
    req.params.file;

  const target =
    path.join(base, filename);

  // ruleid: kisa.sw03.javascript.external-input-to-file-path
  return fs.readFile(target, "utf8", () => {});
}


function unsafeResolveOnly(req, base) {
  const filename =
    req.query.file;

  const target =
    path.resolve(base, filename);

  // resolve만으로는 허용 루트 이탈을 막지 않는다.
  // ruleid: kisa.sw03.javascript.external-input-to-file-path
  return fs.writeFileSync(target, "data");
}


function unsafeMultiStep(req) {
  const raw =
    req.body.path;

  const first =
    raw;

  const second =
    first;

  // ruleid: kisa.sw03.javascript.external-input-to-file-path
  return fs.unlinkSync(second);
}


async function unsafePromises(req) {
  const filename =
    req.query.file;

  // ruleid: kisa.sw03.javascript.external-input-to-file-path
  return fs.promises.readFile(filename, "utf8");
}


function safeBasename(req, base) {
  const filename =
    path.basename(req.query.file);

  const target =
    path.join(base, filename);

  // ok: kisa.sw03.javascript.external-input-to-file-path
  return fs.readFileSync(target, "utf8");
}


function safeHardcoded() {
  // ok: kisa.sw03.javascript.external-input-to-file-path
  return fs.readFileSync(
    "/srv/app/public/help.txt",
    "utf8"
  );
}


function safeUnrelatedStorage(req, storage) {
  const filename =
    req.query.file;

  // ok: kisa.sw03.javascript.external-input-to-file-path
  return storage.readFile(filename);
}

File: backend/semgrep_rules/tests/python/kisa_sw_03_path_traversal.py
import os
from pathlib import Path
import shutil

from flask import send_file
from werkzeug.utils import secure_filename


def unsafe_open(request):
    path = request.GET.get("path")

    # ruleid: kisa.sw03.python.external-input-to-file-path
    return open(path, "r")


def unsafe_resolve_only(request, base):
    raw = request.GET.get("path")

    candidate = (
        Path(base) / raw
    ).resolve()

    # resolve만으로는 base containment가 보장되지 않는다.
    # ruleid: kisa.sw03.python.external-input-to-file-path
    return open(candidate, "r")


def unsafe_multi_step(request):
    raw = request.args.get("path")

    first = raw
    second = first

    # ruleid: kisa.sw03.python.external-input-to-file-path
    return open(second, "rb")


def unsafe_remove(request):
    target = request.POST.get("target")

    # ruleid: kisa.sw03.python.external-input-to-file-path
    os.remove(target)


def unsafe_send_file(request):
    target = request.args.get("file")

    # ruleid: kisa.sw03.python.external-input-to-file-path
    return send_file(target)


def unsafe_copy_destination(request):
    target = request.form.get("output")

    # ruleid: kisa.sw03.python.external-input-to-file-path
    return shutil.copyfile("/srv/app/template.txt", target)


def safe_basename(request, base):
    raw = request.GET.get("path")

    name = os.path.basename(raw)
    target = Path(base) / name

    # ok: kisa.sw03.python.external-input-to-file-path
    return open(target, "r")


def safe_path_name(request, base):
    raw = request.GET.get("path")

    name = Path(raw).name
    target = Path(base) / name

    # ok: kisa.sw03.python.external-input-to-file-path
    return target.read_text()


def safe_secure_filename(request, base):
    raw = request.args.get("file")

    name = secure_filename(raw)
    target = Path(base) / name

    # ok: kisa.sw03.python.external-input-to-file-path
    return open(target, "rb")


def safe_hardcoded():
    # ok: kisa.sw03.python.external-input-to-file-path
    return open(
        "/srv/app/public/help.txt",
        "r",
    )
